multi/stm32l4-multi/uart.c

```diff
@@ -66,308 +46,42 @@ static const int uartConfig[] = { UART1, UART2, UART3, UART4, UART5, LPUART1 };
 		return -EINVAL;
 
-	uart = uartPos[uart];
-
-	if (!bufflen)
-		return 0;
-
-	if (!uart_common[uart].enabled)
-		return -EIO;
-
-	mutexLock(uart_common[uart].txlock);
-	mutexLock(uart_common[uart].lock);
-
-	keepidle(1);
-
-	*(uart_common[uart].base + tdr) = *((unsigned char *)buff);
-	uart_common[uart].txbeg = (void *)((unsigned char *)buff + 1);
-	uart_common[uart].txend = (void *)((unsigned char *)buff + bufflen);
-	*(uart_common[uart].base + cr1) |= 1 << 7;
-
-	while (uart_common[uart].txbeg != uart_common[uart].txend)
-		condWait(uart_common[uart].txcond, uart_common[uart].lock, 0);
-	mutexUnlock(uart_common[uart].lock);
-
-	keepidle(0);
-	mutexUnlock(uart_common[uart].txlock);
-
-	return bufflen;
+	return libuart_write(&uart_common[uartPos[uart]], buff, bufflen);
 }
 
 
 int uart_read(int uart, void* buff, unsigned int count, char mode, unsigned int timeout)
 {
-	int i, err;
-	volatile unsigned int read = 0;
-	char mask = 0x7f;
-
-	if (uart < usart1 || uart > lpuart1 || !uartConfig[uart])
+	if (uart < usart1 || uart > uart5 || !uartConfig[uart])
 		return -EINVAL;
 
-	uart = uartPos[uart];
-
-	if (!uart_common[uart].enabled)
-		return -EIO;
-
-	if (!count)
-		return 0;
-
-	mutexLock(uart_common[uart].rxlock);
-	mutexLock(uart_common[uart].lock);
-
-	uart_common[uart].read = &read;
-	uart_common[uart].rxend = (char *)buff + count;
-
-	/* This field works as trigger for rx interrupt to store data in buffer
-	 * instead of FIFO */
-	uart_common[uart].rxbeg = buff;
-
-	/* Provoke UART exception to fire so that existing data from
-	 * rxdfifo is copied into buff. The handler will clear this
-	 * bit. */
-
-	*(uart_common[uart].base + cr1) |= 1 << 7;
-
-	while (uart_common[uart].rxbeg != uart_common[uart].rxend) {
-		err = condWait(uart_common[uart].rxcond, uart_common[uart].lock, timeout);
-
-		if (mode == uart_mnblock || (timeout && err == -ETIME) || !uart_common[uart].enabled) {
-			uart_common[uart].rxbeg = NULL;
-			uart_common[uart].rxend = NULL;
-			uart_common[uart].read = NULL;
-			break;
-		}
-	}
-
-	if (uart_common[uart].bits < 8) {
-		if (uart_common[uart].bits == 6)
-			mask = 0x3f;
-
-		for (i = 0; i < read; ++i)
-			((char *)buff)[i] &= mask;
-	}
-	mutexUnlock(uart_common[uart].lock);
-
-	mutexUnlock(uart_common[uart].rxlock);
-
-	return read;
+	return libuart_read(&uart_common[uartPos[uart]], buff, count, mode, timeout);
 }
 
 
 int uart_init(void)
 {
 	int i, uart;
-	const struct {
-		volatile uint32_t *base;
-		int dev;
-		unsigned irq;
-	} info[] = {
-		{ (void *)0x40013800, pctl_usart1, usart1_irq },
-		{ (void *)0x40004400, pctl_usart2, usart2_irq },
-		{ (void *)0x40004800, pctl_usart3, usart3_irq },
-		{ (void *)0x40004c00, pctl_uart4, uart4_irq },
-		{ (void *)0x40005000, pctl_uart5, uart5_irq },
-		{ (void *)0x40008000, pctl_lpuart1, lpuart1_irq }
-	};
 
-	for (i = 0, uart = 0; uart < sizeof(info) / sizeof(info[0]); ++uart) {
+	for (i = 0, uart = usart1; uart < uart5; ++uart) {
```

Comment:
You missed initialisation of uart5 this way.

Reply:
Fixed